onnx/defs/reduction/defs.cc

```diff
@@ -36,46 +74,44 @@ False instead of True.)DOC";
         "Constrain input and output types to high-precision numeric tensors.");
     schema.TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
       propagateElemTypeFromInputToOutput(ctx, 0, 0);
-      if (!hasNInputShapes(ctx, 1)) {
-        return;
-      }
-
-      int64_t keep_dims = 1;
-      auto attr_proto = ctx.getAttribute("keepdims");
-      if (attr_proto) {
-        keep_dims = attr_proto->i();
-      }
-      auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
-      int64_t input_ndim = input_shape.dim_size();
-      auto output_shape =
-          ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
-      std::vector<int64_t> axes;
-      auto axes_proto = ctx.getAttribute("axes");
-      if (axes_proto)
-        axes.assign(axes_proto->ints().begin(), axes_proto->ints().end());
-
-      for (size_t i = 0; i < axes.size(); ++i) {
-        if (axes[i] < 0)
-          axes[i] += input_ndim;
-      }
-      // do we need handle negative axis?
-      for (int i = 0; i < input_ndim; ++i) {
-        // axes empty means reduce all dim
-        if (!axes.empty() &&
-            std::find(axes.begin(), axes.end(), i) == axes.end()) {
-          auto dim = output_shape->add_dim();
-          dim->CopyFrom(input_shape.dim(i));
-        } else {
-          if (keep_dims == 1) {
-            auto dim = output_shape->add_dim();
-            dim->set_dim_value(1);
-          }
-        }
-      }
+      reduceShapeInference(ctx);
     });
   };
 }
 
+std::function<void(OpSchema&)> LowPrecisionReduceDocGenerator(
+    const char* name) {
+  return [=](OpSchema& schema) {
+    std::string doc = R"DOC(
+Computes the {name} of the low-precision input tensor's element along the provided axes. The resulted
+tensor has the same rank as the input if keepdims equal 1. If keepdims equal 0, then
+the resulted tensor have the reduced dimension pruned.
+
+The above behavior is similar to numpy, with the exception that numpy default keepdims to
+False instead of True.)DOC";
+    ReplaceAll(doc, "{name}", name);
+    schema.SetDoc(doc.c_str());
+    schema.Attr(
+        "axes",
+        "A list of integers, along which to reduce. The default is to reduce over "
+        "all the dimensions of the input tensor.",
+        AttributeProto::INTS,
+        OPTIONAL);
+    schema.Attr(
+        "keepdims",
+        "Keep the reduced dimension or not, default 1 mean keep reduced dimension.",
+        AttributeProto::INT,
+        static_cast<int64_t>(1));
+    schema.Input(0, "data", "An input tensor.", "T");
+    schema.Output(0, "reduced", "Reduced output tensor.", "tensor(int32)");
+    schema.TypeConstraint(
+        "T",
+        {"tensor(int8)", "tensor(uint8)"},
+        "Constrain input type to 8-bits integer tensors.");
+    schema.TypeAndShapeInferenceFunction(reduceShapeInference);
```

Comment:
Set the output type to be int32

Reply:
output type is already int32 specified in type constraints, don't need to inferred in this function.

Reply:
You do need. Shape/type inference is different with your type constraints (e.g. shape inference engine will put the inferred information into the model as ValueInfoProto).

Reply:
why? type constraints is saying the output type is int32 only, it does not need any type inference. Type inference done by this function is when there're multiple options in output type constraints and no way to infer it with type constraints only. For example, input type is T1 and output type is T2, and T2 could be int32, double. From constraints, it's not doable to get to know which output type should be picked up. In the cases that output type constraint only have one (this case) or input and output types are the same (with type constraints specified both as "T"), type inference function is not needed. Make sense?